src/gosling-brush/linear-brush-model.ts

```diff
@@ -0,0 +1,238 @@
+import * as d3Selection from 'd3-selection';
+import { drag as d3Drag } from 'd3-drag';
+
+const HIDDEN_BRUSH_EDGE_SIZE = 3;
+
+export type oneDimBrushData = [OneDimBrushBodyData, OneDimBrushStartEdgeData, OneDimBrushEndEdgeData];
+
+export type OneDimBrushDataUnion = OneDimBrushBodyData | OneDimBrushStartEdgeData | OneDimBrushEndEdgeData;
+
+export interface OneDimBrushDataCommon {
+    start: number;
+    end: number;
+}
+
+export interface OneDimBrushBodyData extends OneDimBrushDataCommon {
+    type: 'body';
+    cursor: 'grab';
+}
+
+export interface OneDimBrushStartEdgeData extends OneDimBrushDataCommon {
+    type: 'start';
+    cursor: 'ew-resize';
+}
+
+export interface OneDimBrushEndEdgeData extends OneDimBrushDataCommon {
+    type: 'end';
+    cursor: 'ew-resize';
+}
+
+export type OnBrushCallbackFn = (start: number, end: number) => void;
+
+interface BrushStyle {
+    color: string;
+    stroke: string;
+    strokeWidth: number;
+    strokeOpacity: number;
+    opacity: number;
+}
+
+/**
+ * A model to manage 1D brush graphics and its data.
+ */
+export class OneDimBrushModel {
+    /* graphical elements */
+    private brushSelection: d3Selection.Selection<SVGRectElement, OneDimBrushDataUnion, SVGGElement, any>;
+    private readonly style: BrushStyle;
+
+    /* data */
+    private range: [number, number] | null;
+    private data: oneDimBrushData;
+
+    /* drag */
+    private startEvent: typeof d3Selection.event;
+    private prevExtent: [number, number] | null;
+
+    /* visual parameters */
+    private offset: [number, number];
+    private size: number; // fixed size of one-dimension of a brush (e.g., height)
+
+    /* External libraries that we re-use from HiGlass */
+    private externals: {
+        d3Selection: typeof d3Selection;
+        d3Drag: typeof d3Drag;
+    };
+
+    // TODO (May-19-2022): A way to pass only the required function (e.g., onRangeBrush) and allow
+    // to use `this` (i.e., instance of GoslingTrack) in the function?
+    /* gosling track */
+    private track: any;
+
+    constructor(
+        selection: d3Selection.Selection<SVGGElement, unknown, HTMLElement, unknown>,
+        HGC: any,
+        track: any,
+        style: Partial<BrushStyle> = {}
+    ) {
+        this.range = null;
+        this.prevExtent = [0, 0];
+        this.data = this.rangeToData(0, 0);
+
+        this.offset = [0, 0];
+        this.size = 0;
+
+        this.externals = {
+            d3Selection: HGC.d3Selection,
+            d3Drag: HGC.d3Drag.drag
```

Comment:
perhaps,

```suggestion
            d3Selection: HGC.d3Selection,
            d3Drag: HGC.d3Drag
```

for consistency